src/main/java/org/skife/jdbi/v2/sqlobject/BatchHandler.java

```diff
@@ -56,37 +57,17 @@
         this.sql = SqlObject.getSql(anno, raw_method);
         this.transactional = anno.transactional();
         this.batchChunkSize = determineBatchChunkSize(sqlObjectType, raw_method);
-        final GetGeneratedKeys getGeneratedKeys = raw_method.getAnnotation(GetGeneratedKeys.class);
+        this.batchReturner = newBatchReturner(method);
+    }
+
+    private static BatchReturner<?> newBatchReturner(ResolvedMethod method) {
+        GetGeneratedKeys getGeneratedKeys = method.getRawMember().getAnnotation(GetGeneratedKeys.class);
         if (getGeneratedKeys == null) {
-            returner = new Returner()
-            {
-                @Override
-                public int[] value(PreparedBatch batch)
-                {
-                    return batch.execute();
-                }
-            };
-        }
-        else if (getGeneratedKeys.columnName().isEmpty()) {
-            returner = new Returner()
-            {
-                @Override
-                public int[] value(PreparedBatch batch)
-                {
-                    return toPrimitiveArray(batch.executeAndGenerateKeys(IntegerColumnMapper.PRIMITIVE).list());
-                }
-            };
-        }
-        else {
-            returner = new Returner()
-            {
-                @Override
-                public int[] value(PreparedBatch batch)
-                {
-                    String columnName = getGeneratedKeys.columnName();
-                    return toPrimitiveArray(batch.executeAndGenerateKeys(IntegerColumnMapper.PRIMITIVE, columnName).list());
-                }
-            };
+            return BatchReturner.newIntArrayReturner();
+        } else if (method.getReturnType().getErasedType().equals(int[].class)) {
+            return BatchReturner.newGeneratedIntReturner(getGeneratedKeys.columnName());
+        } else {
+            return BatchReturner.newGeneratedLongReturner(getGeneratedKeys.columnName());
```

Comment:
Is it worth checking for `long[]` here and throwing an exception if the return type is unexpected?

Reply:
I'd been thinking about that, but I left it this way to keep it consistent with the code that was here previously. The return type is already checked [here](https://github.com/HubSpot/jdbi/pull/12/files#diff-32537d880332b8f3549a9ac5e82c305bc0a8f70bea0bab1af8308aab505475f5R52-R54).

Reply:
As a follow-up, I just had to update this code to allow `null` (`void`s) as we seem to have cases where people are putting `@GetGeneratedKeys` on `void` methods.

Reply:
I tried adding a test to catch this, but it doesn't seem to NPE in the test suite.